### DISCUSSION TITLE
Filtering pbns by Auction

### DISCUSSION THREAD

**Discussion by Rick-Wilson** [ADavidBailey/Practice-Bidding-Scenarios#42] [Ideas]
The scenario scripts generate hands in /pbn.  Then BBA is run to bid and DD the hands, and these resultant PBNs are placed in /bba.  Not all of the auctions in these hands match the senario goals.  I would like to have a set of pbns, with auctions, which 100% match the goal, in order to auto-generate bidding practice materials.

I have written a Bridge Composer script "Filter.js", which will apply a filter to a pbn file, and save the results to another pbn file.  This script takes in a pbn file, and a regular-expression string, and filters the hands by those regular expressions.  It then saves the output to a new pbn file.  It logs how many hands were scanned, and how many passed the filter.

I've gone through all the scenarios, and created filters for all scenarios supported by BBA's bidding bots.  This amounts to a little over 1/2 of the scenarios.  I've created a Windows script to run the Filter.js on all these files.  I've attached a report showing the results.  I'm sure I will need to fix some of the regular expressions, but also there may be some improvements that can be made to the scenario scripts.
[FilterAll.pdf](https://github.com/user-attachments/files/17249085/FilterAll.pdf)

I think we'll eventually put the filtered pbns into a /filtered folder, as a sibling to /bba.

**Comment by ADavidBailey** [comment]
Nice!  I'm confused about the file names.  The filenames in pbn end with -N, -E, -S, or -W (I need that for my 'rotate' program).  The .pbn files in bba have the bidding added and DO NOT have the dealer appended to the filename.  What's going on?

**Comment by Rick-Wilson** [comment]
I also wrote a BBA script (for the old version that doesn't have AUTOEXIT), and I kept the same input and output filenames, so the BBA folder wound up with the compass suffixes.  Then I write the Filter script, and it also keeps the same names for the input and output files.  So still with the suffix.

I could change that, but I think they will be needed, because we still need rotation on these files too, and it needs to be post-filter.  I think we'll wind up using my Bridge Composer rotation script for this, and it can also use the suffix to determine the proper rotation.

My BC Rotation script also looks for an optional PBN tag ("Rotation_Basis") to direct the rotation.  It might be possible to insert these into the /pbn files, then we wouldn't need the suffix.

**Reply by ADavidBailey** [reply]
I removed the 'dealer suffix' for cosmetic reasons.  Should I leave them?
